src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +340,477 @@ where
+        consolidation::consolidate(&mut self.to_write);
+
+        self.write_to_persist(responders).await?;
+
+        // Wait until our artificial latency has completed.
+        //
+        // Note: if writing to persist took longer than `DEFAULT_TICK` this
+        // await will resolve immediately.
+        tokio::time::sleep_until(min_time_to_complete).await;
+
+        ControlFlow::Continue(())
+    }
+
+    /// Attempt to write what is currently in [Self::to_write] to persist,
+    /// retrying and re-syncing to persist when necessary, that is when the
+    /// upper was not what we expected.
+    async fn write_to_persist(
+        &mut self,
+        responders: Vec<oneshot::Sender<Result<(), StorageError<T>>>>,
+    ) -> ControlFlow<String> {
+        // We'll try really hard to succeed, but eventually stop.
+        //
+        // Note: it's very rare we should ever need to retry, and if we need to
+        // retry it should only take 1 or 2 attempts. We set `max_tries` to be
+        // high though because if we hit some edge case we want to try hard to
+        // commit the data.
+        let retries = Retry::default()
+            .initial_backoff(Duration::from_secs(1))
+            .clamp_backoff(Duration::from_secs(3))
+            .factor(1.25)
+            .max_tries(20)
+            .into_retry_stream();
+        let mut retries = Box::pin(retries);
+
+        loop {
+            // Append updates to persist!
+            let updates_to_write = self
+                .to_write
+                .iter()
+                .map(|(row, diff)| Update {
+                    row: row.clone(),
+                    timestamp: self.current_upper.clone(),
+                    diff: diff.clone(),
+                })
+                .collect();
+
+            let now = T::from((self.now)());
+            let new_upper = std::cmp::max(
+                now,
+                TimestampManipulation::step_forward(&self.current_upper),
+            );
+
+            let request = vec![(
+                self.id,
+                updates_to_write,
+                self.current_upper.clone(),
+                new_upper.clone(),
+            )];
+
+            let append_result = match self.write_handle.compare_and_append(request.clone()).await {
+                // We got a response!
+                Ok(append_result) => append_result,
+                // Failed to receive which means the worker shutdown.
+                Err(_recv_error) => {
+                    // Sender hung up, this seems fine and can happen when
+                    // shutting down.
+                    notify_listeners(responders, || {
+                        Err(StorageError::ShuttingDown("PersistMonotonicWriteWorker"))
+                    });
+
+                    // End the task since we can no longer send writes to persist.
+                    return ControlFlow::Break("sender hung up".to_string());
+                }
+            };
```

Comment:
Not covered according to [coverage/428](https://buildkite.com/materialize/coverage/builds/428).

Reply:
Yeah, same as above. This code path is defensive, but the case that it protects against currently can't really happen. 🙈